src/tactic/norm_cast.lean

```diff
@@ -1,400 +1,648 @@
-do
-    (declaration.thm n l ty e) ← get_decl decl | failed,
-    let tac := λ ty, (flip_eq ty <|> flip_iff ty),
-    (ty', f) ← aux_after_set tac ty,
-    let e' := task.map f e,
-    let n' := new_name n,
-    add_decl (declaration.thm n' l ty' e'),
-    simp_attr.push_cast.set decl () tt
+/--
+`label` is a type used to classify `norm_cast` lemmas.
+* elim lemma:   LHS has 0 head coes and ≥ 1 internal coe,  RHS has 0 coes
+* move lemma:   LHS has 1 head coe and 0 internal coes,    RHS has 0 head coes and ≥ 1 internal coes
+* squash lemma: LHS has ≥ 2 head coes and 0 internal coes, RHS has fewer head coes
+-/
+@[derive [decidable_eq, has_reflect, inhabited]]
+inductive label
+| elim   : label
+| move   : label
+| squash : label
+
+namespace label
+
+/-- Convert `label` into `string`. -/
+protected def to_string : label → string
+| elim   := "elim"
+| move   := "move"
+| squash := "squash"
+
+instance : has_to_string label := ⟨label.to_string⟩
+instance : has_repr label := ⟨label.to_string⟩
+meta instance : has_to_format label := ⟨λ l, l.to_string⟩
+
+/-- Convert `string` into `label`. -/
+def of_string : string -> option label
+| "elim" := some elim
+| "move" := some move
+| "squash" := some squash
+| _ := none
+
+end label
+
+open label
+
+/-- Count how many coercions are at the top of the expression. -/
+meta def count_head_coes : expr → ℕ
+| `(coe %%e) := count_head_coes e + 1
+| `(coe_sort %%e) := count_head_coes e + 1
+| `(coe_fn %%e) := count_head_coes e + 1
+| _ := 0
+
+/-- Count how many coercions are inside the expression, including the top ones. -/
+meta def count_coes : expr → tactic ℕ
+| `(coe %%e) := (+1) <$> count_coes e
+| `(coe_sort %%e) := (+1) <$> count_coes e
+| `(coe_fn %%e) := (+1) <$> count_coes e
+| (app `(coe_fn %%e) x) := (+) <$> count_coes x <*> (+1) <$> count_coes e
+| (expr.lam n bi t e) := do
+  l ← mk_local' n bi t,
+  count_coes $ e.instantiate_var l
+| e := do
+  as ← e.get_simp_args,
+  list.sum <$> as.mmap count_coes
 
-private meta def mk_cache : list name → tactic simp_lemmas :=
-monad.foldl simp_lemmas.add_simp simp_lemmas.mk
+/-- Count how many coercions are inside the expression, excluding the top ones. -/
+private meta def count_internal_coes (e : expr) : tactic ℕ := do
+ncoes ← count_coes e,
+pure $ ncoes - count_head_coes e
 
 /--
-This is an attribute for simplification rules that are
-used to normalize casts.
+aux function for `norm_cast.classify_type`
 
-Let r be = or ↔, then elimination lemmas of the shape
-Π ..., P ↑a1 ... ↑an r P a1 ... an should be given the
-attribute elim_cast.
+remark: the classifier is a little bit less restrictive than the rules describing to the labels.
+This is a consequence of the fact that the rules apply to the explicit coercions appearing in the expression.
+`coe_to_sort` can typically appear in type annotations that are implicit in the displayed expression, but
+will be accounted for in `count_coes`.
```

Comment:
```suggestion
```

This is fixed, right?

Reply:
Yes, the function now computes what the rules say it should.  Although I had to update the description of the rule for squash lemmas from the paper (so that `coe_zero` is squash).

I didn't find any problems or test cases with `coe_to_sort`, so I'm not sure if we've fixed anything.

Another subtle point is that `⇑f x` is not counted as a head coercion.